dynamic/policy_builder.go

```diff
@@ -0,0 +1,237 @@
+package dynamic
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/getyoti/yoti-go-sdk/v3/consts"
+	"github.com/getyoti/yoti-go-sdk/v3/yotierror"
+)
+
+const (
+	authTypeSelfieConst = iota + 1
+	authTypePinConst
+)
+
+// PolicyBuilder constructs a json payload specifying the dynamic policy
+// for a dynamic scenario
+type PolicyBuilder struct {
+	wantedAttributes   map[string]WantedAttribute
+	wantedAuthTypes    map[int]bool
+	isWantedRememberMe bool
+	err                error
+}
+
+// Policy represents a dynamic policy for a share
+type Policy struct {
+	attributes   []WantedAttribute
+	authTypes    []int
+	rememberMeID bool
+}
+
+// WithWantedAttribute adds an attribute from WantedAttributeBuilder to the policy
+func (b *PolicyBuilder) WithWantedAttribute(attribute WantedAttribute) *PolicyBuilder {
+	if b.wantedAttributes == nil {
+		b.wantedAttributes = make(map[string]WantedAttribute)
+	}
+	var key string
+	if attribute.derivation != "" {
+		key = attribute.derivation
+	} else {
+		key = attribute.name
+	}
+	b.wantedAttributes[key] = attribute
+	return b
+}
+
+// WithWantedAttributeByName adds an attribute by its name. This is not the preferred
+// way of adding an attribute - instead use the other methods below
+func (b *PolicyBuilder) WithWantedAttributeByName(name string, options ...interface{}) *PolicyBuilder {
```

Comment:
We do have:
> This is not the preferred way of adding an attribute - instead use the other methods below

on this method. But yes, the problem is still present for the functions which call this one. Aside from Examples, how do you think we can improve the documentation around this? I guess it's done this was to future-proof us if we introduce other options

Reply:
Ruby documents options like this: https://www.rubydoc.info/gems/yoti/1.8.0/Yoti/DynamicSharingService/DynamicPolicyBuilder#with_date_of_birth-instance_method

I suppose we could just add a short explanation to indicate the allowed types, something like this:
```
// WithDateOfBirth adds the date of birth attribute.
// Options accepts multiple SourceConstraint
```

This would need to be duplicated for each method (and updated when new options are introduced), but at least it helps integrators know what the options are.

Reply:
Added in 24a5056 👍